odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,896 @@
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(func.domain, func.domain, linear=False)
+
+            def _call(self, x):
+                """Apply the gradient operator to the given point."""
+                return op.derivative(x).adjoint(func.gradient(op(x)))
+
+        return FunctionalCompositionGradient()
+
+
+class FunctionalRightVectorMult(Functional, OperatorRightVectorMult):
+
+    """Expression type for the functional right vector multiplication.
+
+    Given a functional ``func`` and a vector ``y`` in the domain of ``func``,
+    this corresponds to the functional
+
+        ``(func * y)(x) == func(y * x)``.
+    """
+
+    def __init__(self, func, vector):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The domain of ``func`` must be a ``vector.space``.
+        vector : `domain` element
+            The vector to multiply by.
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('`fun` {!r} is not a `Functional` instance'
+                            ''.format(func))
+
+        OperatorRightVectorMult.__init__(self, operator=func, vector=vector)
+
+        Functional.__init__(self, space=func.domain)
+
+    @property
+    def functional(self):
+        return self.operator
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        return self.vector * self.operator.gradient * self.vector
+
+    @property
+    def convex_conj(self):
+        """Convex conjugate functional of the functional.
+
+        This is only defined for vectors with no zero-elements.
+        """
+        return self.functional.convex_conj * (1.0 / self.vector)
+
+
+class FunctionalSum(Functional, OperatorSum):
+
+    """Expression type for the sum of functionals.
+
+    ``FunctionalSum(func1, func2) == (x --> func1(x) + func2(x))``.
+    """
+
+    def __init__(self, func1, func2):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func1, func2 : `Functional`
+            The summands of the functional sum. Their `Functional.domain`
+            and `Functional.range` must coincide.
+        """
+        if not isinstance(func1, Functional):
+            raise TypeError('`func1` {!r} is not a `Functional` instance'
+                            ''.format(func1))
+        if not isinstance(func2, Functional):
+            raise TypeError('`func2` {!r} is not a `Functional` instance'
+                            ''.format(func2))
+
+        OperatorSum.__init__(self, func1, func2)
+
+        Functional.__init__(self, space=func1.domain,
+                            linear=(func1.is_linear and func2.is_linear),
+                            grad_lipschitz=(func1.grad_lipschitz +
+                                            func2.grad_lipschitz))
+
+    @property
+    def left_func(self):
```

Comment:
Not quite sure if this is needed. `left` does not imply anything about the functional/operator, so users should stick to it.

Having both `left` and `left_func` will cause users to wonder what's up.